dipy/denoise/denspeed.pyx

```diff
@@ -153,49 +157,70 @@ cdef double process_block(double [:, :, ::1] arr,
 
     cdef:
         cnp.npy_intp m, n, o, M, N, O, a, b, c, cnt, step
-        double patch_vol_size
+        double patch_vol_size, block_vol_size
         double summ, d, w, sumw, sum_out, x, sigm
         double * W
         double * cache
         double * sigma_block
         double denom
         cnp.npy_intp BS = B * 2 + 1
+        cnp.npy_intp PS = P * 2 + 1
+        cnp.npy_intp TS = PS + BS - 1
 
     cnt = 0
     sumw = 0
     patch_vol_size = (P + P + 1) * (P + P + 1) * (P + P + 1)
+    block_vol_size = (B + B + 1) * (B + B + 1) * (B + B + 1)
 
-    W = <double *> malloc(BS * BS * BS * sizeof(double))
-    cache = <double *> malloc(BS * BS * BS * sizeof(double))
-    sigma_block = <double *> malloc(BS * BS * BS * sizeof(double))
+    W = <double *> malloc(PS * PS * PS * sizeof(double))
+    cache = <double *> malloc(TS * TS * TS * sizeof(double))
+    sigma_block = <double *> malloc(TS * TS * TS * sizeof(double))
 
     # (i, j, k) coordinates are the center of the static patch
     # copy block in cache
-    copy_block_3d(cache, BS, BS, BS, arr, i - B, j - B, k - B)
-    copy_block_3d(sigma_block, BS, BS, BS, sigma, i - B, j - B, k - B)
+    copy_block_3d(
+        cache,
+        PS + BS - 1,
+        PS + BS - 1,
+        PS + BS - 1,
+        arr,
+        i - P - B,
+        j - P - B,
+        k - P - B)
+    copy_block_3d(
+        sigma_block,
+        PS + BS - 1,
+        PS + BS - 1,
+        PS + BS - 1,
+        arr,
+        i - P - B,
+        j - P - B,
+        k - P - B)
 
     # calculate weights between the central patch and the moving patch in block
     # (m, n, o) coordinates are the center of the moving patch
     # (a, b, c) run inside both patches
-    for m in range(P, BS - P):
-        for n in range(P, BS - P):
-            for o in range(P, BS - P):
```

Comment:
I see that the term in the exponential as defined here (and in the reference you linked) differ from that defined in Eq. 2 of the paper by Coupe et. al.  (http://www.ncbi.nlm.nih.gov/pubmed/18390341).  Namely, both the numerator and denominator of the `exp` exponent are squared in Eq.2 relative to the above.
I think the original Buades paper definition matches the Coupe one.  

This will affect the degree of smoothing as `(a / b) != (sqrt(a) / sqrt(b))` in most cases.  When the numerator of the `exp` term is large relative to `h`, the weights from Coupe should fall off faster than the ones as defined above giving a sharper image.